File: Lab1/ex8.py
from sys import stdin
from math import sqrt

def check_sol(U, V, W, x, y, z):
    return x + y + z == U and x * y * z == V and x ** 2 + y ** 2 + z ** 2 == W and x < y and y < z

T = None
mn = 5 * (10 ** 6)
for line in stdin:
    if T is None:
        T = int(line)
        continue
    else:
        U, V, W = line.split()
    U = int(U)
    V = int(V)
    W = int(W)
    x = None
    y = None
    z = None
    for i in range(1, min(mn, min(V, min(W, U)))):

        tmp_u = U - i
        if V % i != 0:
            continue
        tmp_v = V / i
        if i > U or i * i > W:
            break

        if tmp_u ** 2 < 4 * tmp_v:
            continue
        det = tmp_u ** 2 - 4 * tmp_v
        if int(sqrt(det)) != sqrt(det):
            continue

        root_1 = (tmp_u - sqrt(det)) / 2
        root_2 = (tmp_u + sqrt(det)) / 2

        if int(root_1) == root_1 and root_1 > i:
            k = tmp_u - root_1
            if check_sol(U, V , W, i, root_1, k):
               x = i
               y = root_1
               z = k
               break

        if int(root_2) == root_2 and root_2 > i:
            k = tmp_u - root_2
            if check_sol(U, V, W, i, root_2, k):
               x = i
               y = root_2
               z = k
               break

    if x is None:
        print("empty set")
    else:
        print("{} {} {}".format(int(x), int(y), int(z)))

File: Lab1/ex8.cpp
#include<iostream>
#include<cstdio>
#include<cstdlib>
#include<cstring>
#include<set>
#include<map>
#include<vector>
#include<utility>
#include<queue>
#include<stack>
#include<string>
#include<algorithm>
#include<cmath>

#define ri(X) scanf("%d",&X)
#define rii(X,Y) scanf("%d %d",&X,&Y)
#define rf(X) scanf("%lf",&X)
#define rff(X,Y) scanf("%lf %lf",&X,&Y)
#define mp(X,Y) make_pair(X,Y)
#define pii pair<int,int>
#define FOR(i,j) for(int i=0;i<j;i++)
#define FORC(i,j,c) for(int i=0;i<j;i+=c)

using namespace std;

long long U,V,W, T;
const long long mn = 2.5e6;
const int oo = 1e9;

bool check_sol(long long x, long long y, long long z) {
    return x < y && y < z && x + y + z == U && x * y * z == V && x * x + y * y + z * z == W ;
}
vector< long long > divs;
void preprocess() {
    divs.clear();
    
    double bound = (long long )pow(V, 1.0/3) + 10LL;
    for(int i =1 ; i <  bound; i++) {
        if(V % i == 0) {
            divs.push_back(i);
            divs.push_back(-i);
            divs.push_back(V / i);
            divs.push_back(- V / i);
        }
    }

    sort(divs.begin(), divs.end());

}

int main(){

    scanf("%lld", &T);
    FOR(t,T) {
        scanf("%lld %lld %lld", &U, &V, &W);
        bool found = false;
        long long x,y,z;
        preprocess();
        for(int k =0 ; k < divs.size(); k++) {
            long long i = divs[k];
            if(!i) continue;
            long long tmp_u = U - i;

            long long  tmp_v = V/i;
            if (tmp_u * tmp_u < 4*tmp_v) continue;
            long long det = tmp_u * tmp_u - 4 * tmp_v;
            if(floor(sqrt(det)) != sqrt(det)) continue;

            double root_1 = (tmp_u - sqrt(det)) / 2.0;

            double root_2 = (tmp_u + sqrt(det)) / 2.0;

            if (floor(root_1) == root_1 && root_1 > i) {
                long long k = tmp_u - root_1;
                if (check_sol(i, root_1, k)){
                   x = i;
                   y = root_1;
                   z = k;
                   found= true;
                   break;
                }
            }

            if (floor(root_2) == root_2 && root_2 > i) {
                long long k = tmp_u - root_2;
                if (check_sol(i, root_2, k)) {
                   x = i;
                   y = root_2;
                   z = k;
                   found = true;
                   break;
                }
            }
                        

        }
        if(!found) cout << "empty set" << endl;
        else cout << x <<" " << y << " " << z << endl;

    }

	return 0;
}
